src/Components/src/Microsoft.AspNetCore.Components.Server/Rendering/HtmlRenderer.cs

```diff
@@ -0,0 +1,207 @@
+            var childFrames = childComponentRenderTree.GetFrames();
+            var (segments, _) = RenderFrames(childFrames, 0, childFrames.Count);
+            return (segments, position + frame.ComponentSubtreeLength);
+        }
+
+        private (IEnumerable<string> segments, int nextPosition) RenderElement(ArrayRange<RenderTreeFrame> frames, int position)
+        {
+            var frame = frames.Array[position];
+            IEnumerable<string> elementSegments = new[] { "<", frame.ElementName };
+            var (attributes, afterAttributes) = RenderAttributes(frames, position + 1, frame.ElementSubtreeLength - 1);
+            elementSegments = elementSegments.Concat(attributes);
+            var remainingElements = frame.ElementSubtreeLength + position - afterAttributes;
+            if (remainingElements > 0)
+            {
+                elementSegments = elementSegments.Concat(new[] { ">" });
+                var (children, afterElement) = RenderChildren(frames, afterAttributes, remainingElements);
+                elementSegments = elementSegments.Concat(children).Concat(new[] { "</", frame.ElementName, ">" });
+                Debug.Assert(afterElement == position + frame.ElementSubtreeLength);
+                return (elementSegments, afterElement);
+            }
+            else
+            {
+                elementSegments = elementSegments.Concat(
+                    SelfClosingElements.Contains(frame.ElementName) ?
+                    new[] { " />" } :
+                    new[] { ">", "</", frame.ElementName, ">" });
+                Debug.Assert(afterAttributes == position + frame.ElementSubtreeLength);
+                return (elementSegments, afterAttributes);
+            }
+        }
+
+        private (IEnumerable<string> children, int afterElement) RenderChildren(ArrayRange<RenderTreeFrame> frames, int position, int maxElements)
+        {
+            if (maxElements == 0)
+            {
+                return (Array.Empty<string>(), position);
+            }
+
+            return RenderCore(frames, position, maxElements);
+        }
+
+        private (IEnumerable<string> attributes, int newPosition) RenderAttributes(ArrayRange<RenderTreeFrame> frames, int position, int maxElements)
+        {
+            if (maxElements == 0)
+            {
+                return (Array.Empty<string>(), position);
+            }
+
+            var attributes = new List<string>();
+            for (var i = 0; i < maxElements; i++)
+            {
+                var candidateIndex = position + i;
+                var frame = frames.Array[candidateIndex];
+                if (frame.FrameType != RenderTreeFrameType.Attribute)
+                {
+                    return (attributes, candidateIndex);
+                }
+                if (frame.AttributeValue != null && !ExcludedValueTypes.Any(et => et.IsAssignableFrom(frame.AttributeValue.GetType())))
```

Comment:
I think this can be both simpler and more complicated :)

 * Based on [the logic starting here](https://github.com/aspnet/AspNetCore/blob/master/src/Components/src/Microsoft.AspNetCore.Components.Server/Circuits/RenderBatchWriter.cs#L143), the only two types you need to deal with during HTML rendering are `string` and `bool`. We already convert other value types to strings earlier in the process (inside the RenderTreeBuilder APIs). So instead of having `ExcludedValueTypes`, you can just check whether `frame.AttributeValue` is a `bool`, and if not, use `frame.AttributeValue as string`. So that should simplify things.
 * However, there's a slight extra complication around bool values. If the value is `true`, then we should serialize it as an empty attribute (e.g., `<someElem myAttribute />`). If the value is `false`, then we skip that attribute completely. This is used in cases like `<button disabled=@shouldBeDisabled />`.

Reply:
> * Based on [the logic starting here](https://github.com/aspnet/AspNetCore/blob/master/src/Components/src/Microsoft.AspNetCore.Components.Server/Circuits/RenderBatchWriter.cs#L143), the only two types you need to deal with during HTML rendering are `string` and `bool`. We already convert other value types to strings earlier in the process (inside the RenderTreeBuilder APIs). So instead of having `ExcludedValueTypes`, you can just check whether `frame.AttributeValue` is a `bool`, and if not, use `frame.AttributeValue as string`. So that should simplify things.

Yeah, I wasn't sure of the cases here as Value was of type object, so I went with the ExcludedValueTypes approach.

WRT the handling o bool values that was an oversight on my part.